README.md
# autocross-v1.0
A short text based racing game with a few cars and one track.


changelog.md
# Changelog:
- Added menu file with racing, garage and auto shop options
- Created garage features within menu file
- Created main file where program is run from
- Created race file with corner, start and end functions
- Continuous corner bug fixed by deleting while loop and replacing with 
a set number of corners, each with their own ID and description
- Autoshop created, reduced number of starter cars to 1 (Honda Civic SI)
- Completed vehicle purchasing system, added in-game currency
- Completed part purchasing and installing system
- Fixed bug where vehicle PR does not change when part is installed
- Fixed bug where part is not removed from parts list after purchase
- Cleaned up code, removed excess white space, reduced line lengths, etc.
- Added comments to code


carpgmenu.py
import carpg
import carpg_races
import carpg_autoshop

# Main menu options
mainmenu = [
    'Race',
    'Garage',
    'Autoshop',
    'Quit']

# List of currently available cars
carlist = [
    '2002 Honda Civic SI'
    ]

# List of currently available tracks
tracklist = [
    'Parking Lot'
    ]

# List of autoshop functionalities
options = [
    'Cars',
    'Parts'
    ]

class Menu:
    # Class which grants a main menu functionality
    def __init__(self, enable):

        self.enable = enable

    def main(self):
        # Main menu data
        print(f"\nAUTOCROSS:\n")
        
        for index, value in enumerate(mainmenu, 1):
            print("\t{}. {}".format(index, value))
        
        response = input("\nInput a number to select.\n")
        
        if response == '1':
            print("Race menu has been selected.\n")
            self.race()
        
        elif response == '2':
            print("Garage has been selected.\n")
            self.garage()
        
        elif response == '3':
            print("Autoshop has been selected.\n")
            self.autoshop()
        
        elif response == '4':
            quit()
        
        else:
            print("I didn't understand that.\n")
            self.main()


    def garage(self):
        # Garage menu data, vehicle selection
        print("GARAGE:\n")

        for car in carlist:
            print(car)
        
        response = input("\nWhich car would you like to drive? (inputs must be lowercase, with spaces)\n")
        
        if response == '2002 honda civic si':
            carpg.current = carpg.honda_civic_si_2002
            print(carpg.current.carstats())
            self.main()

        elif response == '1994 mazda miata r':
            if '1994 Mazda Miata R' in carlist:
                carpg.current = carpg.mazda_miata_r_1994
                print(carpg.current.carstats())
                self.main()
            else:
                print("You do not have this car yet\n")
                self.garage()
        
        elif response == '1991 honda crx si':
            if '1991 Honda CRX SI' in carlist:
                carpg.current = carpg.honda_crx_si_1991
                print(carpg.current.carstats())
                self.main()
            else:
                print("You do not have this car yet\n")
                self.garage()
        
        elif response == '2019 volkswagen golf gti':
            if '2019 Volkswagen Golf GTI' in carlist:
                carpg.current = carpg.volkswagen_golf_gti_2019
                print(carpg.current.carstats())
                self.main()
            else:
                print("You do not have this car yet\n")
                self.garage()
        elif response == 'back':
            self.main()
        else:
            print("I didn't understand that.\n")
            self.garage()

    def race(self):
        # Race menu data leading to race start
        print("TRACKS:\n")
        
        for index, value in enumerate(tracklist, 1):
            print("\t{}. {}".format(index, value))
        
        response = input("\nPlease select a track (input number)\n")
        
        if response == '1':
            if carpg.current.pr == 20 or 21:
                carpg_races.pr20.racestart()
            elif carpg.current.pr == 22 or 23:
                carpg_races.pr22.racestart()
            elif carpg.current.pr == 24 or 25:
                carpg_races.pr24.racestart()
            elif carpg.current.pr == 26 or 27:
                carpg_races.pr26.racestart()
            elif carpg.current.pr == 28 or 29:
                carpg_races.pr28.racestart()
            else:
                carpg_races.pr30.racestart()
        elif response == 'back':
            self.main()

        else:
            print("I didn't understand that\n")
            self.race()

    def autoshop(self):
        # Autoshop menu data, current Credits, options leading to auto shops
        print("AUTOSHOP:\n")
        print(f"You currently have {carpg_autoshop.shop.playercr} Credits. You are currently driving your {carpg.current.cardesc()}.\n")
        for index, value in enumerate(options, 1):
            print("\t{}. {}".format(index, value))
        response = input("Which menu would you like to access?\n")
        if response == '1':
            carpg_autoshop.shop.cars_shop()
        elif response == '2':
            carpg_autoshop.shop.parts_install()
        elif response == 'back':
            self.main()
        else:
            print("Invalid input\n")
            self.autoshop()

acmenu = Menu(1)


carpg_races.py
import carpg
import carpgmenu
import carpg_autoshop
import random

class Race:
    # A class that grants the race functionality
    def __init__(self, totaltime):
        self.totaltime = totaltime

    def racestart(self):
        # Start of the race
        if carpg.current == '':
            print("You don't have a car to race with\n")
            carpgmenu.acmenu.main()
        else:
            print ("START!\n")
            self.corner1()

    def corner1(self):
        # Corner 1 data
        response = input("You accelerate through the starting line. There is an easy left corner approaching. Take it fast (risky) or slow (safe)?\n")
        risky_corner = random.randint(1, 6)
        if response == 'safe':
            self.totaltime += 1
            self.corner2()
        elif response == 'risky':
            if risky_corner == 6:
                self.crash()
            else:
                self.totaltime -= 1
                self.corner2()
        else:
            print("Invalid input\n")
            self.corner1()

    def corner2(self):
        # Corner 2 data
        response = input("You exit turn 1. There is a medium right corner approaching. Take it fast (risky) or slow (safe)?\n")
        risky_corner = random.randint(1, 6)
        if response == 'safe':
            self.totaltime += 1
            self.corner3()
        elif response == 'risky':
            if risky_corner == 5 or 6:
                self.crash()
            else:
                self.totaltime -= 1
                self.corner3()
        else:
            print("Invalid input\n")
            self.corner2()

    def corner3(self):
        # Corner 3 data
        response = input("You exit turn 2. There is a 90 degree left corner approaching. Take it fast (risky) or slow (safe)?\n")
        risky_corner = random.randint(1, 6)
        if response == 'safe':
            self.totaltime += 1
            self.corner4()
        elif response == 'risky':
            if risky_corner == 4 or 5 or 6:
                self.crash()
            else:
                self.totaltime -= 1
                self.corner4()
        else:
            print("Invalid input\n")
            self.corner3()

    def corner4(self):
        # Corner 4 data
        response = input("You exit turn 3. There is fast left corner approaching. Take it fast (risky) or slow (safe)?\n")
        risky_corner = random.randint(1, 6)
        if response == 'safe':
            self.totaltime += 1
            self.corner5()
        elif response == 'risky':
            if risky_corner == 6:
                self.crash()
            else:
                self.totaltime -= 1
                self.corner5()
        else:
            print("Invalid input\n")
            self.corner4()

    def corner5(self):
        # Corner 5 data
        response = input("You exit turn 4. There is a narrow hairpin right corner approaching. Take it fast (risky) or slow (safe)?\n")
        risky_corner = random.randint(1, 6)
        if response == 'safe':
            self.totaltime += 1
            self.race_end()
        elif response == 'risky':
            if risky_corner == 3 or 4 or 5 or 6:
                self.crash()
            else:
                self.totaltime -= 1
                self.race_end()
        else:
            print("Invalid input\n")
            self.corner5()

    def race_end(self):
        # End of race data, payments, etc.
        print(f"Your time was {self.totaltime} seconds.\n")
        placement = 1
        opponent1time = random.randint(57, 59)
        opponent2time = random.randint(59, 62)
        if opponent1time <= self.totaltime:
            placement += 1
        elif opponent2time <= self.totaltime:
            placement += 1
        if placement == 1:
            print("You placed 1st. You recieved 1000 Credits.\n")
            carpg_autoshop.shop.playercr += 1000
            carpgmenu.acmenu.main()
        elif placement == 2:
            print("You placed 2nd. You recieved 500 Credits.\n")
            carpg_autoshop.shop.playercr += 500
            carpgmenu.acmenu.main()
        elif placement == 3:
            print("You placed 3rd. You recieved 250 Credits.\n")
            carpg_autoshop.shop.playercr += 250
            carpgmenu.acmenu.main()

    def crash(self):
        # Crash data
        print("You crashed. You recieve a DNF and your car is totalled. You must pay 250 Credits in repairs.\n")
        carpg_autoshop.shop.playercr -= 250
        carpgmenu.acmenu.main()

# Initial race times based on PR of current vehicle
pr20 = Race(60)
pr22 = Race(59.5)
pr24 = Race(59)
pr26 = Race(58.5)
pr28 = Race(58)
pr30 = Race(57.5)
    


carpg_autoshop.py
import carpgmenu
import carpg

# A list of all parts available for purchase and install
partslist = [
    'Engine - $12,000',
    'Transmission - $6000',
    'Brakes - $2000',
    'Exhaust - $500',
    'Suspension - $1000'
    ] 

# A list of all vehicles available for purchase
shopcarslist = [
    '1994 Mazda Miata R - $5100',
    '1991 Honda CRX SI - $6160',
    '2019 Volkswagen Golf GTI - $28,040'
    ]

class Shop:
    # A class which adds the shop functionality of the program
    def __init__(self, playercr):
        self.playercr = playercr
    def cars_shop(self):
        # Function which lets the user purchase additional vehicles
        for shopcar in shopcarslist:
            print(shopcar)
        response = input("Which car would you like to purchase? (input car name, lower case with spaces)\n")
        if response == '1994 mazda miata r':
            # Mazda Miata R
            response = input("Are you sure you would like to buy this car? (yes/no)\n")
            if response == 'yes':
                if self.playercr >= 5100:
                    self.playercr -= 5100
                    carpgmenu.carlist.append('1994 Mazda Miata R')
                    shopcarslist.remove('1994 Mazda Miata R - $5100')
                    print("You've successfully purchased a 1994 Mazda Miata R\n")
                    self.cars_shop()
                else:
                    print("You do not have enough money\n")
                    self.cars_shop()
            elif response == 'no':
                self.cars_shop()
        elif response == '1991 honda crx si':
            # Honda CRX SI
            response = input("Are you sure you would like to buy this car? (yes/no)\n")
            if response == 'yes':
                if self.playercr >= 6160:
                    self.playercr -= 6160
                    carpgmenu.carlist.append('1991 Honda CRX SI')
                    shopcarslist.remove('1991 Honda CRX SI - $6160')
                    print("You've successfully purchased a 1991 Honda CRX SI\n")
                    self.cars_shop()
                else:
                    print("You do not have enough money\n")
                    self.cars_shop()
            elif response == 'no':
                self.cars_shop()
        elif response == '2019 volkswagen golf gti':
            # VW Golf GTI
            response = input("Are you sure you would like to buy this car? (yes/no)\n")
            if response == 'yes':
                if self.playercr >= 28040:
                    self.playercr -= 28040
                    carpgmenu.carlist.append('2019 Volkswagen Golf GTI')
                    shopcarslist.remove('2019 Volkswagen Golf GTI - $28,040')
                    print("You've successfully purchased a 2019 Volkswagen Golf GTI\n")
                    self.cars_shop()
                else:
                    print("You do not have enough money\n")
                    self.cars_shop()
            elif response == 'no':
                self.cars_shop()
        elif response == 'back':
            carpgmenu.acmenu.autoshop()
        else:
            print("Invalid input\n")
            self.cars_shop()

    def parts_install(self):
        # Function which allows the user to purchase parts for their car(s)
        for part in partslist:
            print(part)
        response = input("Select which part you would like to purchase and install. (lowercase)\n")
        if response == 'engine':
            # Engine upgrade
            response = input("Are you sure you want to purchase and install this part? (yes/no)\n")
            if response == 'yes':
                if shop.playercr >= 12000:
                    partslist.remove('Engine - $12,000')
                    carpg.current.hrspr += 50
                    carpg.current.pr += 8
                    print("Part successfully installed.\n")
                    shop.playercr -= 12000
                    self.parts_install()
                else:
                    print("You don't have enough money\n")
                    self.parts_install()
            elif response == 'no':
                self.parts_install()
            else:
                print("Invalid input.\n")
                self.parts_install()
        elif response == 'transmission':
            # Transmission upgrade
            response = input("Are you sure you want to purchase and install this part? (yes/no)\n")
            if response == 'yes':
                if shop.playercr >= 6000:
                    partslist.remove('Transmission - $6000')
                    carpg.current.pr += 4
                    shop.playercr -= 6000
                    print("Part successfully installed.\n")
                    self.parts_install()
                else:
                    print("You don't have enough money\n")
                    self.parts_install()
            elif response == 'no':
                self.parts_install()
            else:
                print("Invalid input.\n")
                self.parts_install()
        elif response == 'brakes':
            # Brakes upgrade
            response = input("Are you sure you want to purchase and install this part? (yes/no)\n")
            if response == 'yes':
                if shop.playercr >= 2000:
                    partslist.remove('Brakes - $2000')
                    shop.playercr -= 2000
                    carpg.current.pr += 2
                    print("Part successfully installed.\n")
                    self.parts_install()
                else:
                    print("You don't have enough money\n")
                    self.parts_install()
            elif response == 'no':
                self.parts_install()
            else:
                print("Invalid input.\n")
                self.parts_install()
        elif response == 'exhaust':
            # Exhaust upgrade
            response = input("Are you sure you want to purchase and install this part? (yes/no)\n")
            if response == 'yes':
                if shop.playercr >= 500:
                    partslist.remove('Exhaust - $500')
                    shop.playercr -= 500
                    carpg.current.hrspr += 5
                    carpg.current.pr += 1
                    print("Part successfully installed.\n")
                    self.parts_install()
                else:
                    print("You don't have enough money\n")
                    self.parts_install()
            elif response == 'no':
                self.parts_install()
            else:
                print("Invalid input.\n")
                self.parts_install()
        elif response == 'suspension':
            # Suspension upgrade
            response = input("Are you sure you want to purchase and install this part? (yes/no)\n")
            if response == 'yes':
                if shop.playercr >= 1000:
                    partslist.remove('Suspension - $1000')
                    shop.playercr -= 1000
                    carpg.current.pr += 2
                    print("Part successfully installed.\n")
                    self.parts_install()
                else:
                    print("You don't have enough money\n")
                    self.parts_install()
            elif response == 'no':
                self.parts_install()
            else:
                print("Invalid input.\n")
                self.parts_install()
        elif response == 'back':
            carpgmenu.acmenu.autoshop()
        else:
            print("Invalid input.\n")
            self.parts_install()

shop = Shop(6000)
            
            
        


carpg.py
class Car:
    # Class which defines all stats and descriptions of a car

    def __init__(self, make, model, year, hrspr, weight, pr):
        self.make = make
        self.model = model
        self.year = year
        self.hrspr = hrspr
        self.weight = weight
        self.pr = pr

    def carstats(self):
        # Car stats complete description
        car_stats = (f"Your {current.year} {current.make} {current.model} weighs {current.weight} lbs" +
        f" and has {current.hrspr} Hrspr. It has a PR of {current.pr}.")
        return car_stats

    def cardesc(self):
        # Complete car description
        car_desc = f"{current.year} {current.make} {current.model}"
        return car_desc

honda_civic_si_2002 = Car('Honda', 'Civic SI', '2002', 160, 2744, 20)
mazda_miata_r_1994 = Car('Mazda', 'Miata R', '1994', 128, 2323, 23)
honda_crx_si_1991 = Car('Honda', 'CRX SI', '1991', 105, 2174, 22)
volkswagen_golf_gti_2019 = Car('Volkswagen', 'Golf GTI', '2019', 228, 3062, 24)

current = honda_civic_si_2002

